primitives/src/smart_extension.rs

```diff
@@ -39,15 +39,71 @@ impl Default for SmartExtensionType {
 pub struct SmartExtensionName(pub Vec<u8>);
 
 #[derive(Encode, Decode, Default, Clone, PartialEq, Debug)]
-/// U type refers to the AccountId which will act
-/// as the address of the smart extension
-pub struct SmartExtension<U> {
+/// Smart Extension details when SE instance
+/// is attached with asset.
+pub struct SmartExtension<AccountId> {
     /// Type of the extension
     pub extension_type: SmartExtensionType,
     /// Name of extension
     pub extension_name: SmartExtensionName,
     /// AccountId of the smart extension
-    pub extension_id: U,
+    pub extension_id: AccountId,
     /// Status of the smart extension
     pub is_archive: bool,
 }
+
+/// The url string of the SE template.
+#[derive(Encode, Decode, Default, Debug, Clone, PartialEq, VecU8StrongTyped)]
+pub struct MetaUrl(pub Vec<u8>);
+
+/// The description string about the SE template.
+#[derive(Encode, Decode, Default, Debug, Clone, PartialEq, VecU8StrongTyped)]
+pub struct MetaDescription(pub Vec<u8>);
+
+/// The version string of the SE template.
+#[derive(Encode, Decode, Default, Debug, Clone, PartialEq, VecU8StrongTyped)]
+pub struct MetaVersion(pub Vec<u8>);
+
+/// Subset of the SE template metadata that is provided by the template owner.
+#[derive(Encode, Decode, Default, Debug, Clone, PartialEq)]
+pub struct SmartExtensionMetadata<Balance> {
+    /// Url that can contain the details about the template
+    /// Ex- license, audit report.
+    pub url: Option<MetaUrl>,
+    /// Type of smart extension template.
+    pub se_type: SmartExtensionType,
+    /// Fee paid at the time on creating new instance form the template.
+    pub instantiation_fee: Balance,
+    /// Fee paid at the time of usage of the SE (A given operation performed).
+    pub usage_fee: Balance,
```

Comment:
Fees should be stored separately from meta details because we need to read fees onchain. We don't need to read other meta details when we only need the fees.

Reply:
I know that this PR only contains the very basic framework to record usage fees, but IMO a good approach would be to also store a copy of the usage fee associated with each individual instance (i.e. take the usage fee from the template and copy it to the instance at the point of instantiation). This means a user will get a fixed usage fee rather than it being possible for the author to modify it after someone has decided to use an extension.

Reply:
I will take care of the fee changes in the next PR.